packages/amagaki-plugin-partial-library/src/partial-library.ts

```diff
@@ -0,0 +1,356 @@
+    options?: PartialLibraryPluginOptions
+  ): PartialLibraryPluginOptions | undefined {
+    pod.plugins.register(PartialLibraryPlugin, options);
+
+    return options;
+  }
+  partials: Record<string, PartialLibraryPartial> = {};
+
+  constructor(public pod: Pod, public config: PartialLibraryPluginOptions) {}
+
+  addPartialInstance(
+    key: string,
+    partialConfig: Record<string, any>,
+    fileName?: string
+  ) {
+    if (!(key in this.partials)) {
+      this.partials[key] = new PartialLibraryPartial(key);
+    }
+
+    // Ignore the ignored partials.
+    if (this.config.partial?.ignored?.includes(key)) {
+      return;
+    }
+
+    // Track only (no config) for the tracked partials.
+    if (this.config.partial?.tracked?.includes(key)) {
+      this.partials[key].addInstance(undefined, fileName);
+      return;
+    }
+
+    this.partials[key].addInstance(partialConfig, fileName);
+  }
+
+  /**
+   * Hook for finding all of the partial usage when rendering.
+   */
+  async afterRenderHook(result: TemplateEngineRenderResult): Promise<void> {
+    const docPartials =
+      result.context.doc?.fields[this.config.document?.key ?? 'partials'] ?? [];
+
+    for (const partialConfig of docPartials) {
+      const partialKey = this.config.partial?.key ?? 'partial';
+      const partial = partialConfig[partialKey];
+      this.addPartialInstance(partial, partialConfig, result.path);
+    }
+  }
+
+  /**
+   * Hook for generating the library files before the manifest is generated.
+   *
+   * This allows for all of the document partials to be parsed and tracked
+   * before generating the library output and still be included in the
+   * build manifest for deployment.
+   *
+   * @param builder Builder instance
+   */
+  async beforeBuildManifestHook(
+    builder: Builder,
+    createdPaths: Array<CreatedPath>,
+    artifacts: Array<Artifact>
+  ): Promise<void> {
+    const partialKeys = Object.keys(this.partials).sort();
+    const bar = Builder.createProgressBar(
+      this.config?.build?.loadingLabel ?? 'Partial library'
+    );
+    const startTime = new Date().getTime();
+
+    // Pages for all partials plus main library index.
+    bar.start(partialKeys.length + 1, 0, {
+      customDuration: Builder.formatProgressBarTime(0),
+    });
+
+    const pathPrefix = this.config.serving?.pathPrefix ?? '/library/';
+
+    // Add the main library page.
+    const normalPath = Builder.normalizePath(pathPrefix);
+    const tempPath = fsPath.join(
+      builder.tempDirRoot,
+      builder.outputDirectoryPodPath,
+      normalPath
+    );
+    const realPath = this.pod.getAbsoluteFilePath(
+      fsPath.join(builder.outputDirectoryPodPath, normalPath)
+    );
+
+    const createdPath = {
+      tempPath,
+      normalPath,
+      realPath,
+    };
+
+    // TODO: Use a nunjucks template.
+    const timer = this.pod.profiler.timer(
+      'library.render',
+      'Partial Library render'
+    );
+
+    try {
+      const context = {
```

Comment:
Instead of adding a new hook to add files to the build manifest, is there a reason this can't (or shouldn't) just be added as routes? If they are added as routes then they serve in the dev server and are built alongside the static build.

Reply:
The issue with adding them as normal routes is that we need the entire route tree to be built before we have all the information to generate the library files. As far as I know we don't have a way to say "run these routes last".